pkg/agent/monitortool/monitor.go

```diff
@@ -161,30 +197,118 @@ func (m *MonitorTool) pingAll() {
 	}
 }
 
-func (m *MonitorTool) pingNode(ctx context.Context, ip string) (bool, time.Duration) {
-	pinger, err := ping.NewPinger(ip)
+// resolveIP resolves the target address to an IP address.
+func resolveIPV4(ctx context.Context, target string) (*net.IPAddr, error) {
+	IPProtocol := "ip4"
+
+	klog.InfoS("Resolving target address", "ip_protocol", IPProtocol)
+
+	resolver := &net.Resolver{}
+	ips, err := resolver.LookupIPAddr(ctx, target)
+	if err != nil {
+		klog.ErrorS(err, "Resolution with IP protocol failed")
+		return nil, err
+	}
+
+	// Return the first IPv4 address
+	for _, ip := range ips {
+		if ip.IP.To4() != nil {
+			klog.InfoS("Resolved target address", "ip", ip.String())
+			return &ip, nil
+		}
+	}
+
+	return nil, fmt.Errorf("unable to find IPv4 address for target")
+}
+
+func (m *MonitorTool) pingNode(ctx context.Context, addr string) (bool, time.Duration) {
+	// Resolve the IP address
+	ip, err := resolveIPV4(ctx, addr)
+	if err != nil {
+		klog.ErrorS(err, "Failed to resolve IP address")
+
+		return false, 0
+	}
+
+	srcIP := net.ParseIP("0.0.0.0")
+	requestType := ipv4.ICMPTypeEcho
+	replyType := ipv4.ICMPTypeEchoReply
+
+	// TODO: Enable DontFragment?
+	socket, err := net.ListenPacket("ip4:icmp", srcIP.String())
 	if err != nil {
-		// Try to show log
+		klog.ErrorS(err, "Failed to listen on ICMP")
 		return false, 0
 	}
-	defer pinger.Stop()
+	defer socket.Close()
 
-	// Get timeout channel
-	go func() {
-		<-ctx.Done()
-		pinger.Stop()
-	}()
+	// Create a new ICMP packet
+	body := &icmp.Echo{
+		ID:   os.Getpid() & 0xffff, // Use the current process ID as the ICMP ID
+		Seq:  int(getICMPSeq()),
+		Data: []byte("HELLO-ANTREA-AGENT"),
+	}
+	msg := icmp.Message{
+		Type: requestType,
+		Code: 0,
+		Body: body,
+	}
 
-	err = pinger.Run()
+	// Serialize the ICMP message
+	msgBytes, err := msg.Marshal(nil)
 	if err != nil {
-		// Try to show log
+		klog.ErrorS(err, "Failed to marshal ICMP message")
+		return false, 0
+	}
+	rttStart := time.Now()
+	if _, err = socket.WriteTo(msgBytes, ip); err != nil {
```

Comment:
something that I have been wondering:
- does the antrea-agent Pod have all the necessary capabilities for this (i.e., is NET_ADMIN sufficient: https://github.com/antrea-io/antrea/blob/7415b938022e0feb44d6ef4bb2878c93053d2540/build/yamls/antrea.yml#L6969)
- will this work for antrea-agent on Windows Nodes?

Reply:
I have not tested it, but it works well in current config.

Reply:
If we do not have admin permisson, we need to use `udp` instead.